### ISSUE TITLE
ITV episode only partial downloading then fails

### ISSUE THREAD

**Issue by mikeytt** [Ascoware/get-iplayer-automator#67]
*What steps will reproduce the problem? Be specific, and provide as much detail as possible.*
Select ITV show in Safari, process PID, press start. Starts to download then stops at random stages and fails to complete

*What is the expected output? What do you see instead?*
Successful Download; Download Failed

*What version of GiA are you using? On what version of OS X?*
Version 1.9.7 (20171007004) 
macOS Yosemite 10.10.5

*Are you using a web proxy? DNS proxy? VPN? VPS?*
No to all

*What program are you attempting to download (full title, series and episode)? **Provide only ONE program as an example.** *
X factor series 14 episode 14
https://www.itv.com/hub/the-x-factor/1a8647a0264

*Please provide additional information*
Tried on three different machines, tried multiple formats
[RTMPDump verbose.txt](https://github.com/Ascoware/get-iplayer-automator/files/1387275/RTMPDump.verbose.txt)

**Comment by nitramlatep** [comment]
I have the same issue but just with this program, so wondering if its an issue with the ITV Hub file itself?  I have been able to watch it on the ITV hub however directly.

**Comment by aeromaxx** [comment]
What steps will reproduce the problem? Be specific, and provide as much detail as possible.
Select ITV show in Chrome, Use Current Webpage, press start. Starts to download then stops at random stages and fails to complete.

What is the expected output? What do you see instead?
Successful Download; Download Failed

What version of GiA are you using? On what version of OS X?
Version 1.9.7 (20171007004) 
macOS El Capitan 10.11.6

Are you using a web proxy? DNS proxy? VPN? VPS?
No to all

*What program are you attempting to download (full title, series and episode)? Provide only ONE program as an example. *
X factor series 14 episode 14
https://www.itv.com/hub/the-x-factor/1a8647a0264

Please provide additional information
This programme has only 28 days left until it can no longer be downloaded, I'm not sure if the information that has been given already will be enough to resolve whatever the issue is after the programme can no longer be downloaded and the fix to be tested.

[GetI_iPlayer_Automator_Verbose_Log.txt](https://github.com/Ascoware/get-iplayer-automator/files/1389909/GetI_iPlayer_Automator_Verbose_Log.txt)

**Comment by Vangelis66** [comment]
@nitramlatep wrote: 

> I have been able to watch it on the ITV hub however directly. 

... Apples and oranges; ITVHub in a desktop browser (e.g. Google Chrome & Firefox) currently uses encrypted MPEG-DASH streams, in mobile browsers AppleHLS streams; GiA still uses the older Flash infrastructure, which delivers unencrypted (non-DRMed) RTMPE streams... 

@mikeytt & @aeromaxx; 

You are both in the UK, first is trying to fetch the 896x504p@1500kbps quality variant (1-8647-0264-001_PC011500_16X9.mp4), while the second the best one, 896x504p@1800kbps (1-8647-0264-001_PC011800_16X9.mp4); both on GiA 1.9.7 (20171007004) , with the same RTMPDump binary (v2.4-102-ga3a600d-get_iplayer). 

I am currently away from a Mac, but out of pure curiosity I tried today to fetch the same ITV programme using [youtube-dl](https://github.com/rg3/youtube-dl/releases/) in a Win7SP1 32bit laptop (as there doesn't exist a Windows equivalent of GiA :sob: ); this Python script also uses RTMPDump to fetch ITV VOD (FFmpeg is also used in an auxiliary fashion to remux to the MP4 container and add basic metadata, even subtitles if need be); I am finding that the programme downloads OK: 
```
youtube-dl -F "https://www.itv.com/hub/the-x-factor/1a8647a0264"
[ITV] 1a8647a0264: Downloading webpage
[ITV] 1a8647a0264: Downloading XML
[info] Available formats for 1a8647a0264:
format code  extension  resolution note
rtmp-600     flv        unknown     600k
rtmp-800     flv        unknown     800k
rtmp-1200    flv        unknown    1200k
rtmp-1500    flv        unknown    1500k
rtmp-1800    flv        unknown    1800k  (best)
```
`youtube-dl "https://www.itv.com/hub/the-x-factor/1a8647a0264" -o "The X Factor - Se14Ep14[1a8647a0264].mp4" --add-metadata` => 
```
[ITV] 1a8647a0264: Downloading webpage
[ITV] 1a8647a0264: Downloading XML
[download] Destination: The X Factor - Se14Ep14[1a8647a0264].mp4
[download]  99.9% of ~1009.67MiB at  1.17MiB/s ETA 00:00
[rtmpdump] 1057658831 bytes
[download] 100% of 1008.66MiB
[ffmpeg] Adding metadata to 'The X Factor - Se14Ep14[1a8647a0264].mp4'
```
So _today_, it doesn't appear as though it's a problem on the ITV CDN's side; nor does it look like an issue with actual GiA code, AFAICT; I would suggest both of you remove any partial downloads and retry that fetch at a time of day your connection is most robust; if you could also grab a different RTMPDump binary (or compile it yourselves), it would not hurt trying with that one, too... 

As you both say, this appears to be an issue isolated to that specific show, this fact alone makes troubleshooting all the more difficult... :disappointed: 

Hope I've helped, albeit a little...

**Comment by nitramlatep** [comment]
@Vangelis66 thanks so much.  Tried it just now, no set up changes, roughly same time of day and worked just fine.  Must of been something in the file posted by ITV - and to that point, on Sunday late night when I first tried to download the program, it was actually on the ITV Hub as a 60min program.  This must of changed at some point to 120min on Monday afternoon so I suspected then that something was wrong at the source.  

If @mikeytt can confirm it works as well then I think this case can be closed.

Also would like to thank @skovatch for what you have been doing maintaining this.  It has been a life line for me and the children more several years.
